core/src/compression/stream.cc

```diff
@@ -27,6 +29,106 @@
 using namespace com::centreon::broker;
 using namespace com::centreon::broker::compression;
 
+/**
+ * Compression function
+ *
+ * @param data the data to compress.
+ * @param compression_level The compression level, by default -1.
+ *
+ * @return The same data compressed.
+ */
+static QByteArray _compress(QByteArray const& data, int compression_level) {
+  if (data.isEmpty())
+    return QByteArray(4, '\0');
+
+  uLongf nbytes = (uLongf)data.size();
+
+  if (compression_level < -1 || compression_level > 9)
+    compression_level = -1;
+
+  uLongf len = compressBound(nbytes);
+  QByteArray retval;
+  int res;
+  do {
+    retval.resize(len + 4);
+    res = ::compress2(
+        (Bytef*)retval.data() + 4,
+        &len,
+        (Bytef const*)data.constData(),
+        nbytes,
+        compression_level);
+
+    switch (res) {
+      case Z_OK:
+        retval.resize(len + 4);
+        retval[0] = (nbytes >> 24) & 0xff;
+        retval[1] = (nbytes >> 16) & 0xff;
+        retval[2] = (nbytes >> 8) & 0xff;
+        retval[3] = (nbytes & 0xff);
+        break;
+      case Z_MEM_ERROR:
+        throw (exceptions::msg()
+          << "compress: Z_MEM_ERROR: Not enough memory");
+        break;
+      case Z_BUF_ERROR:
+        len <<= 1;
+        break;
+    }
+  } while (res == Z_BUF_ERROR);
+
+  return retval;
+}
+
+/**
+ * Uncompress function
+ *
+ * @param data The data to extract.
+ * @param nbytes The data size in bytes.
+ *
+ * @return the extract data
+ */
+static QByteArray _uncompress(const uchar* data, uLong nbytes) {
+  if (!data) {
+    logging::debug(logging::medium)
+      << "uncompress: data is null";
+    return QByteArray();
+  }
+  if (nbytes <= 4) {
+    if (nbytes < 4 || (data[0] != 0 || data[1] != 0
+                   || data[2] != 0 || data[3] != 0))
+      logging::debug(logging::medium)
+        << "uncompress: data is corrupted";
+    return QByteArray();
+  }
+  ulong expected_size = (data[0] << 24) | (data[1] << 16)
+                     | (data[2] <<  8) | data[3];
+  ulong len = qMax(expected_size, 1ul);
+  if (len > (ulong)std::numeric_limits<int>::max())
+    throw (exceptions::msg()
+      << "uncompress: data expected size is too big.");
+  QByteArray uncompressed_array(len, '\0');
+
+  ulong alloc = len;
+
+  int res = ::uncompress((Bytef*)uncompressed_array.data(), &len,
+      (Bytef const*)data + 4, nbytes - 4);
+
+  switch (res) {
+    case Z_OK:
+      if (len != alloc)
+        uncompressed_array.resize(len);
+      return uncompressed_array;
+    case Z_MEM_ERROR:
+      throw (exceptions::msg()
+        << "uncompress: Z_MEM_ERROR: not enough memory.");
```

Comment:
"compression: not enough memory to uncompress " << nbytes " compressed bytes to " << len << " uncompressed bytes";